pkg/controller/servicebinding/secret_test.go

```diff
@@ -38,17 +36,98 @@ func TestSecretNew(t *testing.T) {
 		f.FakeDynClient(),
 		ns,
 		name,
+		data,
+		secretOwnerReference,
 	)
 
+	t.Run("unique_secret_name", func(t *testing.T) {
+		hashedName, err := HashedResourceNameForData(name, data)
+		assert.NoError(t, err)
+		assert.Contains(t, hashedName, name)
+	})
+
 	t.Run("createOrUpdate", func(t *testing.T) {
-		u, err := s.createOrUpdate(data, secretOwnerReference)
+		u, err := s.createOrUpdate()
 		assert.NoError(t, err)
+		hashedName, err := HashedResourceNameForData(name, data)
+		assert.NoError(t, err)
+		assert.Equal(t, hashedName, u.GetName())
 		assertSecretNamespacedName(t, u, ns, name)
 	})
 
 	t.Run("Get", func(t *testing.T) {
 		u, err := s.get()
 		assert.NoError(t, err)
+		hashedName, err := HashedResourceNameForData(name, data)
+		assert.NoError(t, err)
+		assert.Equal(t, hashedName, u.GetName())
 		assertSecretNamespacedName(t, u, ns, name)
 	})
+
+	t.Run("different_secret_payload_have_different_name", func(t *testing.T) {
+		payload1 := map[string][]byte{"dbName": []byte("postgres")}
+		value1 := newSecret(
+			f.FakeDynClient(),
+			ns,
+			name,
+			payload1,
+			secretOwnerReference,
+		)
+		secret1, err := value1.createOrUpdate()
+		assert.NoError(t, err)
+
+		payload2 := map[string][]byte{"user": []byte("admin")}
+		value2 := newSecret(
+			f.FakeDynClient(),
+			ns,
+			name,
+			payload2,
+			secretOwnerReference,
+		)
+		secret2, err := value2.createOrUpdate()
+		assert.NoError(t, err)
+
+		assert.NotEqual(t, secret1.GetName(), secret2.GetName())
+	})
+
+	t.Run("secret_with_empty_payload", func(t *testing.T) {
+		value1 := newSecret(
+			f.FakeDynClient(),
+			ns,
+			name,
+			nil,
+			secretOwnerReference,
+		)
+		secret1, err := value1.createOrUpdate()
+		assert.NoError(t, err)
+		assert.NotEqual(t, name, secret1.GetName())
+	})
+
+	t.Run("secret_exists", func(t *testing.T) {
```

Comment:
What is being tested here? First time it'll create the resource, the second time it'll be a no-op since the payload hasn't been changed so neither `Update` or `Create` are executed; perhaps `createOrUpdate` could be renamed to `overwriteSecret(ns, name string, payload map[string][]byte, ownerRef corev1.OwnerReference, force bool)` to add clarity to what it really does.

Reply:
> What is being tested here?

The expectation is that if secret with the same name exists, compare its content and return it.

>  the second time it'll be a no-op since the payload hasn't been changed so neither Update or Create are executed;

Well, that is what I need help with. I am not able to write a unit test for the Update(..) scenario.

> createOrUpdate could be renamed to overwriteSecret(ns, name string, payload map[string][]byte, ownerRef corev1.OwnerReference, force bool) to add clarity to what it really does.

No because:
[1] I don't like the name `overwriteSecret`. `createOrUpdate` is not overwriting the secret if it already exists.
[2] Why do we need a `force` option?

Reply:
> > What is being tested here?
> 
> The expectation is that if secret with the same name exists, compare its content and return it if no difference.
> 
> > the second time it'll be a no-op since the payload hasn't been changed so neither Update or Create are executed;
> 
> Well, that is what I need help with. I am not able to write a unit test for the Update(..) scenario.
> This line -> https://github.com/redhat-developer/service-binding-operator/pull/750/files#diff-0141e5e3383464862fa2299425f3318875159d89221a6cd2f36088f6d06d75b1R111
> 
> How do I create a scenario where `s.name == existingSecret.name` but `s.data != existingSecret's data`

You need to stage the scenario:

1. Create a new reconciler test
1. Add a Deployment to the Fake client to act as application
1. Add a Postgres (or ConfigMap) to the Fake client to act as service
1. Add a Secret to the Fake client the same way the Operator would do (with the contents of the Postgres or ConfigMap variables, named accordingly)
1. Add a SBR resource in the Fake client referring to both Deployment and Postgres (or ConfigMap), and the Secret in the Status
1. Execute `reconciler.Reconcile(...)`
1. Verify whether the Secret has been fixed by the Operator.

> > createOrUpdate could be renamed to overwriteSecret(ns, name string, payload map[string][]byte, ownerRef corev1.OwnerReference, force bool) to add clarity to what it really does.
> 
> No because:
> [1] `overwriteSecret` name doesnot add clarity. `createOrUpdate` is not overwriting the secret if it already exists.

It would if the semantics are changed to always overwrite if the expected payload in the resource is not found.

> [2] Why do we need a `force` option?

We don't, but a `dry-run` of sorts might be desirable.

Reply:
> You need to stage the scenario:

There is already a unit test for this. 
https://github.com/redhat-developer/service-binding-operator/pull/750/files#diff-bc9b850e2f5d5d0c79ab5588cf3713eb9c4765b2f1c562bfc8d92a01e16887f0R155

> It would if the semantics are changed to always overwrite if the expected payload in the resource is not found.

Done